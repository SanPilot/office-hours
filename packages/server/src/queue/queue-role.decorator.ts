import { createParamDecorator, ExecutionContext } from '@nestjs/common';
import { UserModel } from 'profile/user.entity';
import { QueueModel } from './queue.entity';

export const QueueRole = createParamDecorator(
  async (data: unknown, ctx: ExecutionContext) => {
    const request = ctx.switchToHttp().getRequest();
    const queue = await QueueModel.findOne(request.params.queueId);
    const courseId = queue?.courseId;
    const user = await UserModel.findOne(request.user.userId, {
      relations: ['courses'],
    });

    const userCourse = user.courses.find((course) => {
      return Number(course.courseId) === Number(courseId);
    });
    return userCourse.role;
  },
);
