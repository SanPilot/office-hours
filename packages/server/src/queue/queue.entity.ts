import { OpenQuestionStatus } from '@template/common';
import { Exclude, Expose } from 'class-transformer';
import {
  BaseEntity,
  Column,
  Entity,
  JoinColumn,
  JoinTable,
  ManyToMany,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { CourseModel } from '../course/course.entity';
import { UserModel } from '../profile/user.entity';
import { QuestionModel } from '../question/question.entity';

@Entity('queue_model')
export class QueueModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => CourseModel, (course) => course.queues)
  @JoinColumn({ name: 'courseId' })
  course: CourseModel;

  @Column({ nullable: true })
  @Exclude()
  courseId: number;

  @Column('text')
  room: string;

  @OneToMany((type) => QuestionModel, (qm) => qm.queue)
  @Exclude()
  questions: QuestionModel[];

  @Column('text', { nullable: true })
  notes: string;

  @ManyToMany((type) => UserModel, (user) => user.queues)
  @JoinTable()
  staffList: UserModel[];

  time?: {
    start: Date;
    end: Date;
  };

  @Expose()
  get queueSize(): number {
    return (
      this.questions?.filter((q) => q.status in OpenQuestionStatus).length || 0
    );
  }
  // TODO: eventually figure out how staff get sent to FE as well
}
