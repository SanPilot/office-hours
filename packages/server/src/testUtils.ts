import { createConnection, Connection } from "../../ormconfig";
import { init } from "./server";
import path from "path";
import { Server } from "@hapi/hapi";

export function setupServerTest(): () => Server {
  let server: Server;

  beforeEach(async () => {
    server = await init();
  });

  afterEach(async () => {
    await server.stop();
  });

  return () => server;
}

export function setupDBTest() {
  let connection: Connection;
  beforeAll(async () => {
    connection = await createConnection({
      type: "postgres",
      url: "postgres://postgres@localhost:5432/test",
      synchronize: true,
      logging: false,
      entities: [path.resolve(__dirname, "entity/**/*{.js,.ts}")],
      migrations: [path.resolve(__dirname, "migration/**/*{.js,.ts}")],
      subscribers: [path.resolve(__dirname, "subscriber/**/*{.js,.ts}")],
    });
  });

  beforeEach(async () => {
    await connection.synchronize(true);
  });

  afterAll(async () => {
    await connection.close();
  });
}
