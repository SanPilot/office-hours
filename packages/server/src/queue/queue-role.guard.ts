import { Injectable } from '@nestjs/common';
import { UserModel } from '../profile/user.entity';
import { RolesGuard } from 'guards/role.guard';
import { QueueModel } from './queue.entity';

@Injectable()
export class QueueRolesGuard extends RolesGuard {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  async setupData(
    request: any,
  ): Promise<{ courseId: number; user: UserModel }> {
    const queue = await QueueModel.findOne(request.params.queueId);
    const courseId = queue.course.id;
    const user = request.user;
    return { courseId, user };
  }
}
