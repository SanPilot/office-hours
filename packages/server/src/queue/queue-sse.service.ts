import { Injectable } from '@nestjs/common';
import { SSEService } from 'sse/sse.service';
import { QueueService } from './queue.service';
import { Response } from 'express';
import { throttle } from 'lodash';
import { Role } from '@template/common';

type QueueClientMetadata = { userId: number; role: Role };

const idToRoom = (queueId: number) => `q-${queueId}`;
/**
 * Handle sending queue sse events
 */
@Injectable()
export class QueueSSEService {
  constructor(
    private queueService: QueueService,
    private sseService: SSEService<QueueClientMetadata>,
  ) {}

  subscribeClient(
    queueId: number,
    res: Response,
    metadata: QueueClientMetadata,
  ): void {
    this.sseService.subscribeClient(idToRoom(queueId), { res, metadata });
  }

  // Send event with new questions, but no more than once a second
  updateQuestions = throttle(
    async (queueId: number) => {
      try {
        if (queueId) {
          const questions = await this.queueService.getQuestions(queueId);
          if (questions) {
            this.sseService.sendEvent(idToRoom(queueId), ({ role, userId }) => {
              return {
                questions: this.queueService.anonymizeQuestions(
                  questions,
                  userId,
                  role,
                ),
              };
            });
          }
        }
      } catch (e) {}
    },
    1000,
    {
      leading: false,
      trailing: true,
    },
  );

  updateQueue = throttle(
    async (queueId: number) => {
      try {
        const queue = await this.queueService.getQueue(queueId);
        if (queue) {
          this.sseService.sendEvent(idToRoom(queueId), () => {
            return { queue: queue };
          });
        }
      } catch (e) {}
    },
    1000,
    {
      leading: false,
      trailing: true,
    },
  );
}
