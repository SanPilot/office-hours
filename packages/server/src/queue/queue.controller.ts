import {
  Controller,
  Get,
  Patch,
  Param,
  NotFoundException,
  ClassSerializerInterceptor,
  UseInterceptors,
  UseGuards,
  Body,
} from '@nestjs/common';
import { Connection } from 'typeorm';
import { Not, In } from 'typeorm';
import { QueueModel } from './queue.entity';
import {
  GetQueueResponse,
  ListQuestionsResponse,
  ClosedQuestionStatus,
  UpdateQueueNotesParams,
} from '@template/common';
import { QuestionModel } from '../question/question.entity';
import { JwtAuthGuard } from '../profile/jwt-auth.guard';

@Controller('queues')
@UseGuards(JwtAuthGuard)
@UseInterceptors(ClassSerializerInterceptor)
export class QueueController {
  constructor(private connection: Connection) {}

  @Get(':queueId')
  async getQueue(@Param('queueId') queueId: string): Promise<GetQueueResponse> {
    return QueueModel.findOne(queueId, { relations: ['questions'] });
  }

  @Get(':queueId/questions')
  async getQuestions(
    @Param('queueId') queueId: string,
  ): Promise<ListQuestionsResponse> {
    // todo: need a way to return different data, if TA vs. student hits endpoint.
    // for now, just return the student response
    const queueSize = await QueueModel.count({
      where: { id: queueId },
    });
    // Check that the queue exists
    if (queueSize === 0) {
      throw new NotFoundException();
    }

    return await QuestionModel.find({
      where: [
        {
          queueId: queueId,
          status: Not(In(Object.values(ClosedQuestionStatus))),
        },
      ],
      relations: ['creator', 'taHelped'],
    });
  }

  @Patch(':queueId')
  async updateQueue(
    @Param('queueId') queueId: number,
    @Body() body: UpdateQueueNotesParams,
    // TODO: Add TA/Prof protection on endpoint
  ) {
    let queue = await QueueModel.findOne({
      where: { id: queueId },
    });
    if (queue === undefined) {
      throw new NotFoundException();
    }

    queue.notes = body.notes;
    await queue.save();
    return queue;
  }
}
