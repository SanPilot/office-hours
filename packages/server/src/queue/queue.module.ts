import { Module } from '@nestjs/common';
import { QueueController } from './queue.controller';
import { QueueService } from './queue.service';
import { SSEModule } from 'sse/sse.module';

@Module({
  controllers: [QueueController],
  providers: [QueueService],
  exports: [QueueService],
  imports: [SSEModule],
})
export class QueueModule {}
