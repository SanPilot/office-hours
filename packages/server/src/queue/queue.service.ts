import { Injectable } from '@nestjs/common';
import { Cron, CronExpression } from '@nestjs/schedule';
import { Connection } from 'typeorm';
import { QueueModel } from './queue.entity';
import { OpenQuestionStatus } from '@template/common';
import { QuestionModel } from '../question/question.entity';

@Injectable()
export class QueueService {
  constructor(private connection: Connection) {}

  @Cron(CronExpression.EVERY_HOUR)
  public async cleanQueue(queueId: number): Promise<void> {
    const queue = await QueueModel.findOne(queueId, {
      relations: ['staffList', 'questions'],
    });
    // If no staff are present, clean the queue
    if (!queue.staffList || queue.staffList.length === 0) {
      const openQuestions = queue.questions.filter((q) =>
        Object.values(OpenQuestionStatus),
      );
      await Promise.all(
        openQuestions.map(async (q: QuestionModel) => {
          q.status = 'Stale';
          await q.save();
        }),
      );
    }
  }
}
