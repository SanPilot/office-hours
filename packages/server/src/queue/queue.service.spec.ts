import { Test, TestingModule } from '@nestjs/testing';
import { TestTypeOrmModule } from '../../test/util/testUtils';
import { QueueService } from './queue.service';
import {
  QueueFactory,
  UserFactory,
  QuestionFactory,
} from '../../test/util/factories';
import { OpenQuestionStatus } from '@template/common';
import { QuestionModel } from '../question/question.entity';

describe('QueueService', () => {
  let service: QueueService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [TestTypeOrmModule],
      providers: [QueueService],
    }).compile();
    service = module.get<QueueService>(QueueService);
  });

  describe('cleanQueue', () => {
    it('queue remains the same if any staff are checked in', async () => {
      const ta = await UserFactory.create();
      const queue = await QueueFactory.create({ staffList: [ta] });
      await QuestionFactory.create({
        status: OpenQuestionStatus.Queued,
        queue: queue,
      });

      await service.cleanQueue(queue.id);
      const question = await QuestionModel.findOne({});
      expect(question.status).toEqual('Queued');
    });

    it('if no staff are present all questions with open status are marked as stale', async () => {
      const queue = await QueueFactory.create();
      const question = await QuestionFactory.create({
        status: OpenQuestionStatus.Queued,
        queue: queue,
      });

      await service.cleanQueue(queue.id);
      const updatedQuestion = await QuestionModel.findOne(question.id);
      expect(updatedQuestion.status).toEqual('Stale');
    });
  });
});
