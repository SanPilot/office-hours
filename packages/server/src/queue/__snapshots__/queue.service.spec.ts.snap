// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueueService personalizeQuestions hides details of other students 1`] = `
ListQuestionsResponse {
  "priorityQueue": Array [],
  "questionsGettingHelp": Array [],
  "queue": Array [
    QuestionModel {
      "closedAt": null,
      "createdAt": 2020-11-02T12:00:00.000Z,
      "creator": UserModel {
        "email": "user@neu.edu",
        "firstName": "User",
        "id": 1,
        "lastName": null,
        "name": "User",
        "photoURL": "https://pics/user",
      },
      "helpedAt": null,
      "id": 1,
      "isOnline": null,
      "location": null,
      "questionType": "Other",
      "status": "Queued",
      "taHelped": null,
      "text": "help us",
    },
    QuestionModel {
      "closedAt": null,
      "createdAt": 2020-11-02T12:00:00.000Z,
      "creator": UserModel {
        "id": 2,
      },
      "helpedAt": null,
      "id": 2,
      "isOnline": null,
      "location": null,
      "questionType": "Other",
      "status": "Queued",
      "taHelped": null,
      "text": "help someone else",
    },
  ],
  "yourQuestion": QuestionModel {
    "closedAt": null,
    "createdAt": 2020-11-02T12:00:00.000Z,
    "creator": UserModel {
      "desktopNotifsEnabled": false,
      "email": "user@neu.edu",
      "firstName": "User",
      "id": 1,
      "insights": null,
      "lastName": null,
      "name": "User",
      "phoneNotifsEnabled": false,
      "photoURL": "https://pics/user",
    },
    "creatorId": 1,
    "firstHelpedAt": null,
    "helpedAt": null,
    "id": 1,
    "isOnline": null,
    "location": null,
    "questionType": "Other",
    "queueId": 1,
    "status": "Queued",
    "taHelped": null,
    "taHelpedId": null,
    "text": "help us",
  },
}
`;
